# Wine Project

### Index
1.  [Dataset Features Description](#1)<br>
2.  [Load and Discover Data](#2)<br>



</div>

## 1: Dataset Features Description<a id="1"></a>

1. Fixed Acidity

Determines the tartness and crispness of the wine. Higher acidity often means a more refreshing, crisp taste.
White Wines: Generally preferred to be slightly higher for freshness and preservation.
Red Wines: Balanced acidity is crucial for good structure and aging potential.

2. Volatile Acidity

Contributes to the wine’s aroma but in excess can lead to a vinegary smell, indicating spoilage.
White Wines: Lower levels are preferred to avoid spoilage and maintain freshness.
Red Wines: Can tolerate slightly higher levels, but excessive amounts are still undesirable.

3. Citric Acid

Adds a citrusy flavor and helps in preserving the wine. Can enhance freshness and balance.
White Wines: Commonly used to boost freshness; higher levels are often more acceptable.
Red Wines: Less common and used in smaller amounts, as red wines rely more on other acids for structure.

4. Residual Sugar

Determines the sweetness of the wine. Higher residual sugar results in sweeter wines.
White Wines: Can vary widely (dry to sweet). Sweetness must match the style of the wine (e.g., dessert wines have high residual sugar).
Red Wines: Usually drier, but sweeter styles exist (e.g., Port). Excessive sweetness can mask other flavors.

5. Chlorides

Impacts the wine's taste and mouthfeel. High chloride levels can lead to a salty or off taste.
Both: Should be kept within normal ranges to avoid undesirable flavors and ensure quality.

6. Sulfur Dioxide

Acts as a preservative to prevent oxidation and spoilage. Can also impact the wine’s aroma and taste.
White Wines: Generally limited to around 200 mg/L. Essential for freshness and preventing spoilage.
Red Wines: Can tolerate slightly higher levels (up to 250 mg/L) due to their composition and aging process.

7. Density

Reflects the wine’s alcohol content and sugar levels. Denser wines often have higher alcohol or sugar.
Both: Density helps in assessing wine style and balance. Typical ranges are used to ensure consistency and quality.

8. pH

Influences the wine's stability, taste, and aging potential. Lower pH means higher acidity and more freshness.
White Wines: Preferred pH is between 2.9 and 3.6 for crispness and preservation.
Red Wines: Typically between 3.2 and 3.8, affecting structure and aging potential.

9. Sulphates

Adds to the wine’s flavor profile and preservation. Can impact mouthfeel and stability.
Both: Should be balanced to enhance flavor without overpowering the wine.

10. Alcohol

Impacts the body, sweetness, and overall balance of the wine. Higher alcohol can lead to a fuller body and potentially sweeter taste.
Both: Should align with the wine style. Excessive alcohol can overshadow other flavors, while too little can result in a thin wine.

## 2: Load Data<a id="2"></a>

In [1]:
# Import Required Libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

plt.style.use('ggplot')
pd.set_option('display.max_columns', 200) # to display all rows

In [2]:
import chardet
import csv

def detect_file_properties(file_path):
    # Step 1: Check the encoding of the dataset
    with open(file_path, 'rb') as file:
        rawdata = file.read(10000)
        encoding_result = chardet.detect(rawdata)

    # Step 2: Check the separator (delimiter)
    with open(file_path, 'r', encoding=encoding_result['encoding']) as file:
        sample = file.read(10000)
        sniffer = csv.Sniffer()
        has_header = sniffer.has_header(sample)
        dialect = sniffer.sniff(sample)

    # Return the encoding, delimiter, and header info as a dictionary
    return {
        "encoding": encoding_result['encoding'],
        "confidence": encoding_result['confidence'],  # Optional: confidence in encoding detection
        "separator": dialect.delimiter,
        "has_header": has_header
    }




In [3]:

detect_file_properties(r'C:\Users\User\Documents\Python Scripts\Wine-Quality\winequality-red.csv')



{'encoding': 'ascii', 'confidence': 1.0, 'separator': ';', 'has_header': True}

In [4]:
# Load the data 

path_red = r'C:\Users\User\Documents\Python Scripts\Wine-Quality\winequality-red.csv'

path_white = r'C:\Users\User\Documents\Python Scripts\Wine-Quality\winequality-white.csv'


df_red = pd.read_csv(path_red, sep=';', header=0, encoding='ascii').copy()

df_white = pd.read_csv(path_white, sep=';', header=0, encoding='ascii').copy()

In [5]:
df_red.shape

(1599, 12)

In [6]:
df_white.shape

(4898, 12)

In [7]:
df_white.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality  
0      8.8        6  
1      9.5        6  
2     10.1        6  
3      9.9        6  
4      9.9        6

In [8]:
df_red.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [9]:
# Rename multiple columns
df_red.rename(columns={
    'fixed acidity': 'fixed_acidity',
    'volatile acidity': 'volatile_acidity',
    'citric acid': 'citric_acid',
    'residual sugar': 'residual_sugar',
    'free sulfur dioxide': 'free_sulfur_dioxide',
    'total sulfur dioxide': 'total_sulfur_dioxide'
}, inplace=True)


In [10]:
df_red.head(2)

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70          0.0             1.9      0.076   
1            7.8              0.88          0.0             2.6      0.098   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   

   alcohol  quality  
0      9.4        5  
1      9.8        5

In [11]:
# Rename multiple columns
df_white.rename(columns={
    'fixed acidity': 'fixed_acidity',
    'volatile acidity': 'volatile_acidity',
    'citric acid': 'citric_acid',
    'residual sugar': 'residual_sugar',
    'free sulfur dioxide': 'free_sulfur_dioxide',
    'total sulfur dioxide': 'total_sulfur_dioxide'
}, inplace=True)

In [12]:
# Add column for wine type

df_red.loc[:, "type"] = "red"

df_white.loc[:, "type"] = "white"

In [13]:
df_white.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality   type  
0      8.8        6  white  
1      9.5        6  white  
2     10.1        6  white  
3      9.9        6  white  
4      9.9        6  white

### Discover dataset

In [14]:
df_red.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed_acidity         1599 non-null   float64
 1   volatile_acidity      1599 non-null   float64
 2   citric_acid           1599 non-null   float64
 3   residual_sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free_sulfur_dioxide   1599 non-null   float64
 6   total_sulfur_dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
 12  type                  1599 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 162.5+ KB


In [15]:
df_white.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed_acidity         4898 non-null   float64
 1   volatile_acidity      4898 non-null   float64
 2   citric_acid           4898 non-null   float64
 3   residual_sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free_sulfur_dioxide   4898 non-null   float64
 6   total_sulfur_dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
 12  type                  4898 non-null   object 
dtypes: float64(11), int64(1), object(1)
memory usage: 497.6+ KB


In [16]:
df_red.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality type  
0      9.4        5  red  
1      9.8        5  red  
2      9.8        5  red  
3      9.8        6  red  
4      9.4        5  red

In [17]:
df_white.head()

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
0            7.0              0.27         0.36            20.7      0.045   
1            6.3              0.30         0.34             1.6      0.049   
2            8.1              0.28         0.40             6.9      0.050   
3            7.2              0.23         0.32             8.5      0.058   
4            7.2              0.23         0.32             8.5      0.058   

   free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
0                 45.0                 170.0   1.0010  3.00       0.45   
1                 14.0                 132.0   0.9940  3.30       0.49   
2                 30.0                  97.0   0.9951  3.26       0.44   
3                 47.0                 186.0   0.9956  3.19       0.40   
4                 47.0                 186.0   0.9956  3.19       0.40   

   alcohol  quality   type  
0      8.8        6  white  
1      9.5        6  white  
2     10.1        6  white  
3      9.9        6  white  
4      9.9        6  white

In [18]:
df_red.dtypes

fixed_acidity           float64
volatile_acidity        float64
citric_acid             float64
residual_sugar          float64
chlorides               float64
free_sulfur_dioxide     float64
total_sulfur_dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
type                     object
dtype: object

In [19]:
df_white.dtypes

fixed_acidity           float64
volatile_acidity        float64
citric_acid             float64
residual_sugar          float64
chlorides               float64
free_sulfur_dioxide     float64
total_sulfur_dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
type                     object
dtype: object

In [20]:
df_red.isna().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
type                    0
dtype: int64

In [21]:
df_white.isna().sum()

fixed_acidity           0
volatile_acidity        0
citric_acid             0
residual_sugar          0
chlorides               0
free_sulfur_dioxide     0
total_sulfur_dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
type                    0
dtype: int64

In [22]:
df_red.duplicated().sum()

240

In [23]:
# check for duplicates 

red_duplicates = df_red.duplicated()

df_red_duplicates = df_red[red_duplicates]
df_red_duplicates


fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
4               7.4             0.700         0.00            1.90      0.076   
11              7.5             0.500         0.36            6.10      0.071   
27              7.9             0.430         0.21            1.60      0.106   
40              7.3             0.450         0.36            5.90      0.074   
65              7.2             0.725         0.05            4.65      0.086   
...             ...               ...          ...             ...        ...   
1563            7.2             0.695         0.13            2.00      0.076   
1564            7.2             0.695         0.13            2.00      0.076   
1567            7.2             0.695         0.13            2.00      0.076   
1581            6.2             0.560         0.09            1.70      0.053   
1596            6.3             0.510         0.13            2.30      0.076   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
4                    11.0                  34.0  0.99780  3.51       0.56   
11                   17.0                 102.0  0.99780  3.35       0.80   
27                   10.0                  37.0  0.99660  3.17       0.91   
40                   12.0                  87.0  0.99780  3.33       0.83   
65                    4.0                  11.0  0.99620  3.41       0.39   
...                   ...                   ...      ...   ...        ...   
1563                 12.0                  20.0  0.99546  3.29       0.54   
1564                 12.0                  20.0  0.99546  3.29       0.54   
1567                 12.0                  20.0  0.99546  3.29       0.54   
1581                 24.0                  32.0  0.99402  3.54       0.60   
1596                 29.0                  40.0  0.99574  3.42       0.75   

      alcohol  quality type  
4         9.4        5  red  
11       10.5        5  red  
27        9.5        5  red  
40       10.5        5  red  
65       10.9        5  red  
...       ...      ...  ...  
1563     10.1        5  red  
1564     10.1        5  red  
1567     10.1        5  red  
1581     11.3        5  red  
1596     11.0        6  red  

[240 rows x 13 columns]

In [24]:
df_red = df_red.drop_duplicates()

In [25]:
df_red.shape

(1359, 13)

In [26]:
df_white.duplicated().sum()

937

In [27]:
# check  duplicates 

white_duplicates = df_white.duplicated()

df_white_duplicates = df_white[white_duplicates]
df_white_duplicates.head(50).T

4       5      7      8       20      39      48   \
fixed_acidity            7.2     8.1    7.0    6.3     6.2     7.3     7.0   
volatile_acidity        0.23    0.28   0.27    0.3    0.66    0.24    0.31   
citric_acid             0.32     0.4   0.36   0.34    0.48    0.39    0.26   
residual_sugar           8.5     6.9   20.7    1.6     1.2   17.95     7.4   
chlorides              0.058    0.05  0.045  0.049   0.029   0.057   0.069   
free_sulfur_dioxide     47.0    30.0   45.0   14.0    29.0    45.0    28.0   
total_sulfur_dioxide   186.0    97.0  170.0  132.0    75.0   149.0   160.0   
density               0.9956  0.9951  1.001  0.994  0.9892  0.9999  0.9954   
pH                      3.19    3.26    3.0    3.3    3.33    3.21    3.13   
sulphates                0.4    0.44   0.45   0.49    0.39    0.36    0.46   
alcohol                  9.9    10.1    8.8    9.5    12.8     8.6     9.8   
quality                    6       6      6      6       8       5       6   
type                   white   white  white  white   white   white   white   

                         61      63      74      75      88      89      90   \
fixed_acidity            6.0     6.6     6.7     7.4     7.1     7.1     7.2   
volatile_acidity        0.19    0.38    0.23    0.24    0.43    0.44    0.39   
citric_acid             0.26    0.15    0.31    0.29    0.61    0.62    0.63   
residual_sugar          12.4     4.6     2.1    10.1    11.8    11.8    11.0   
chlorides              0.048   0.044   0.046    0.05   0.045   0.044   0.044   
free_sulfur_dioxide     50.0    25.0    30.0    21.0    54.0    52.0    55.0   
total_sulfur_dioxide   147.0    78.0    96.0   105.0   155.0   152.0   156.0   
density               0.9972  0.9931  0.9926  0.9962  0.9974  0.9975  0.9974   
pH                       3.3    3.11    3.33    3.13    3.11    3.12    3.09   
sulphates               0.36    0.38    0.64    0.35    0.45    0.46    0.44   
alcohol                  8.9    10.2    10.7     9.5     8.7     8.7     8.7   
quality                    6       6       8       5       5       6       6   
type                   white   white   white   white   white   white   white   

                         99      104     106    108     114     118     131  \
fixed_acidity            6.0     7.4     7.1    7.1     6.7     7.2     6.5   
volatile_acidity        0.34    0.25    0.12   0.23    0.41    0.31    0.24   
citric_acid             0.66    0.37    0.32   0.35    0.34     0.5    0.32   
residual_sugar          15.9    13.5     9.6   16.5     9.2    13.3     7.6   
chlorides              0.046    0.06   0.054   0.04   0.049   0.056   0.038   
free_sulfur_dioxide     26.0    52.0    64.0   60.0    29.0    68.0    48.0   
total_sulfur_dioxide   164.0   192.0   162.0  171.0   150.0   195.0   203.0   
density               0.9979  0.9975  0.9962  0.999  0.9968  0.9982  0.9958   
pH                      3.14     3.0     3.4   3.16    3.22    3.01    3.45   
sulphates                0.5    0.44    0.41   0.59    0.51    0.47    0.54   
alcohol                  8.8     9.1     9.4    9.1     9.1     9.2     9.7   
quality                    6       5       5      6       5       5       7   
type                   white   white   white  white   white   white   white   

                         141    152     156     159     160     163    191  \
fixed_acidity            7.1    6.9     7.4     5.2     6.1     7.4    6.8   
volatile_acidity        0.37   0.25    0.24    0.44    0.32    0.24   0.28   
citric_acid             0.67    0.3    0.42    0.04    0.24    0.42    0.4   
residual_sugar          10.5    4.1    14.0     1.4     1.5    14.0   22.0   
chlorides              0.045  0.054   0.066   0.036   0.036   0.066  0.048   
free_sulfur_dioxide     49.0   23.0    48.0    43.0    38.0    48.0   48.0   
total_sulfur_dioxide   155.0  116.0   198.0   119.0   124.0   198.0  167.0   
density               0.9975  0.994  0.9979  0.9894  0.9898  0.9979  1.001   
pH

In [28]:
df_white = df_white.drop_duplicates()

In [29]:
df_white.shape

(3961, 13)

In [30]:
df_red.describe(include= 'all')

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  \
count     1359.000000       1359.000000  1359.000000     1359.000000   
unique            NaN               NaN          NaN             NaN   
top               NaN               NaN          NaN             NaN   
freq              NaN               NaN          NaN             NaN   
mean         8.310596          0.529478     0.272333        2.523400   
std          1.736990          0.183031     0.195537        1.352314   
min          4.600000          0.120000     0.000000        0.900000   
25%          7.100000          0.390000     0.090000        1.900000   
50%          7.900000          0.520000     0.260000        2.200000   
75%          9.200000          0.640000     0.430000        2.600000   
max         15.900000          1.580000     1.000000       15.500000   

          chlorides  free_sulfur_dioxide  total_sulfur_dioxide      density  \
count   1359.000000          1359.000000           1359.000000  1359.000000   
unique          NaN                  NaN                   NaN          NaN   
top             NaN                  NaN                   NaN          NaN   
freq            NaN                  NaN                   NaN          NaN   
mean       0.088124            15.893304             46.825975     0.996709   
std        0.049377            10.447270             33.408946     0.001869   
min        0.012000             1.000000              6.000000     0.990070   
25%        0.070000             7.000000             22.000000     0.995600   
50%        0.079000            14.000000             38.000000     0.996700   
75%        0.091000            21.000000             63.000000     0.997820   
max        0.611000            72.000000            289.000000     1.003690   

                 pH    sulphates      alcohol      quality  type  
count   1359.000000  1359.000000  1359.000000  1359.000000  1359  
unique          NaN          NaN          NaN          NaN     1  
top             NaN          NaN          NaN          NaN   red  
freq            NaN          NaN          NaN          NaN  1359  
mean       3.309787     0.658705    10.432315     5.623252   NaN  
std        0.155036     0.170667     1.082065     0.823578   NaN  
min        2.740000     0.330000     8.400000     3.000000   NaN  
25%        3.210000     0.550000     9.500000     5.000000   NaN  
50%        3.310000     0.620000    10.200000     6.000000   NaN  
75%        3.400000     0.730000    11.100000     6.000000   NaN  
max        4.010000     2.000000    14.900000     8.000000   NaN

In [31]:
df_white.describe(include= 'all')

fixed_acidity  volatile_acidity  citric_acid  residual_sugar  \
count     3961.000000       3961.000000  3961.000000     3961.000000   
unique            NaN               NaN          NaN             NaN   
top               NaN               NaN          NaN             NaN   
freq              NaN               NaN          NaN             NaN   
mean         6.839346          0.280538     0.334332        5.914819   
std          0.866860          0.103437     0.122446        4.861646   
min          3.800000          0.080000     0.000000        0.600000   
25%          6.300000          0.210000     0.270000        1.600000   
50%          6.800000          0.260000     0.320000        4.700000   
75%          7.300000          0.330000     0.390000        8.900000   
max         14.200000          1.100000     1.660000       65.800000   

          chlorides  free_sulfur_dioxide  total_sulfur_dioxide      density  \
count   3961.000000          3961.000000           3961.000000  3961.000000   
unique          NaN                  NaN                   NaN          NaN   
top             NaN                  NaN                   NaN          NaN   
freq            NaN                  NaN                   NaN          NaN   
mean       0.045905            34.889169            137.193512     0.993790   
std        0.023103            17.210021             43.129065     0.002905   
min        0.009000             2.000000              9.000000     0.987110   
25%        0.035000            23.000000            106.000000     0.991620   
50%        0.042000            33.000000            133.000000     0.993500   
75%        0.050000            45.000000            166.000000     0.995710   
max        0.346000           289.000000            440.000000     1.038980   

                 pH    sulphates      alcohol      quality   type  
count   3961.000000  3961.000000  3961.000000  3961.000000   3961  
unique          NaN          NaN          NaN          NaN      1  
top             NaN          NaN          NaN          NaN  white  
freq            NaN          NaN          NaN          NaN   3961  
mean       3.195458     0.490351    10.589358     5.854835    NaN  
std        0.151546     0.113523     1.217076     0.890683    NaN  
min        2.720000     0.220000     8.000000     3.000000    NaN  
25%        3.090000     0.410000     9.500000     5.000000    NaN  
50%        3.180000     0.480000    10.400000     6.000000    NaN  
75%        3.290000     0.550000    11.400000     6.000000    NaN  
max        3.820000     1.080000    14.200000     9.000000    NaN

* Wide ranges and extrem max values appear for both types which means we have outliers for many features. 
* I will refer on EU Laws to review the regulations and limits for the features. 
* Website: https://european-union.europa.eu/index_en
* Some features might have higher or lower than standard values which may effect the quality and taste but we will keep these records as long as they follow regulations.

* For Red Wines:

* Sulfur Dioxide (SO2): The maximum total sulfur dioxide limit is 150 mg/L. However, for red wines with a residual sugar content of at least 5 g/L, this can increase to 250 mg/L. 

* Volatile Acidity: It is not universally defined but the general threshold for volatile acidity (acetic acid) is around 1.2 g/L for red wines. Higher means, they can have spoilage or production issues.

* Rest of the Features can greatly influence red wine quality but they are not tightly regulated. 

* White Wines: 

* Maximum total sulfur dioxide limit is 200 mg/L.


#### Review the records which exceeds legal limits: